File: backend/calculations/loan_schedule.py
"""
calculations/loan_schedule.py — Term Loan amortisation (CA standard)

BUG 7 FIX: Use half-yearly reducing balance schedule:
  HalfYearlyInstalment = TermLoan / RepaymentHalfYears
  Interest_HalfYearN   = BalanceAtStartOfHalfYear × (interestRate / 2)
  PrincipalRepaid_HalfYearN = HalfYearlyInstalment (0 during moratorium)
  ClosingBalance_HalfYearN  = BalanceAtStartOfHalfYear − PrincipalRepaid_HalfYearN

BUG FIX (moratorium granularity): moratorium used to be converted to whole
years via math.ceil(moratorium_months / 12) — a 6-month moratorium was
silently ROUNDED UP to a full 12-month moratorium (doubling it), while every
display in the report kept showing the original "6 Month(s)" the applicant
actually entered. Moratorium is now tracked in HALF-YEAR units (rounded to
the nearest half-year, since repayment already happens on a half-yearly
cycle), so a 6-month moratorium correctly skips principal for exactly one
half-year, not a full year.
"""
import math
from core.engine import R

_ZERO_YEAR = lambda y: {
    "year": y, "opening_balance": 0.0, "emi_paid": 0.0,
    "interest_paid": 0.0, "principal_paid": 0.0, "closing_balance": 0.0,
    "half_yearly_instalment": 0.0,
    "mid_year_balance": 0.0, "interest_h1": 0.0, "interest_h2": 0.0,
}


def calculate_loan_schedule(data, scheme_data: dict) -> list:
    """
    Build a 5-year term loan repayment schedule (half-yearly reducing balance).
    Always returns exactly 5 rows.
    """
    term_loan        = float(scheme_data.get("term_loan", 0) or 0)
    rate_pct         = float(getattr(data.assumptions, "interest_rate_pct", 10.5) or 10.5)
    tenure_mo        = int(getattr(data.assumptions, "tenure_months", 60) or 60)
    moratorium_mo    = int(getattr(data.assumptions, "moratorium_months", 0) or 0)

    if term_loan <= 0 or tenure_mo <= 0:
        return [_ZERO_YEAR(y) for y in range(1, 6)]

    annual_rate         = rate_pct / 100
    half_rate           = annual_rate / 2
    tenure_years        = math.ceil(tenure_mo / 12)
    total_half_years    = tenure_years * 2
    # Nearest half-year unit — repayment already runs on a half-yearly cycle,
    # so this is the finest granularity a moratorium can be honoured at.
    moratorium_half_yrs = min(round(moratorium_mo / 6), total_half_years)
    repay_half_years    = max(total_half_years - moratorium_half_yrs, 1)

    half_inst = R(term_loan / repay_half_years, 2)

    rows    = []
    balance = term_loan

    for yr in range(1, tenure_years + 1):
        opening = R(balance, 2)
        bal = opening

        # Half-year 1 of this year
        hy1 = 2 * yr - 1
        ih1 = R(bal * half_rate, 2)
        repaid_h1 = 0.0 if hy1 <= moratorium_half_yrs else min(half_inst, bal)
        bal = R(max(bal - repaid_h1, 0), 2)

        # Half-year 2 of this year
        hy2 = 2 * yr
        ih2 = R(bal * half_rate, 2)
        repaid_h2 = 0.0 if hy2 <= moratorium_half_yrs else min(half_inst, bal)
        closing = R(max(bal - repaid_h2, 0), 2)

        repaid    = R(repaid_h1 + repaid_h2, 2)
        total_int = R(ih1 + ih2, 2)
        emi_paid  = R(repaid + total_int, 2)

        rows.append({
            "year":                   yr,
            "opening_balance":        opening,
            "emi_paid":               emi_paid,
            "interest_paid":          total_int,
            "principal_paid":         repaid,
            "closing_balance":        closing,
            "half_yearly_instalment": half_inst,
            # The TRUE balance after H1 / before H2 — NOT the arithmetic
            # mean of opening and closing. In a year that transitions out
            # of moratorium mid-year (H1 still in moratorium, H2 repaying),
            # the balance is unchanged after H1, so averaging opening and
            # closing would understate it.
            "mid_year_balance":       bal,
            "interest_h1":            ih1,
            "interest_h2":            ih2,
        })
        balance = closing

    # Pad to minimum 5 rows so income statement (5-year projection) always has data
    while len(rows) < 5:
        rows.append(_ZERO_YEAR(len(rows) + 1))
    # Return all rows (not capped to 5) — Section H displays the full schedule
    return rows

File: src/lib/loanSchedule.ts
/**
 * loanSchedule.ts — Term Loan amortisation, mirroring
 * backend/calculations/loan_schedule.py EXACTLY (half-yearly equal-principal,
 * reducing balance — NOT the EMI formula). The backend has never used EMI
 * for the fixed-capital term loan; several frontend spots independently
 * approximated interest/principal with their own ad-hoc formula (the
 * standard EMI formula, a flat-interest EMI-lookalike, or a "x0.9" reducing-
 * balance guess) — each producing a DIFFERENT, backend-disagreeing number.
 * This is the ONE place that should be used everywhere a Year-1 (or any
 * year's) interest/principal split is needed on the frontend.
 */

export interface LoanScheduleYear {
  year: number;
  openingBalance: number;
  interestPaid: number;
  principalPaid: number;
  closingBalance: number;
  halfYearlyInstalment: number;
}

const Rs2 = (n: number) => Math.round(n * 100) / 100;

/**
 * Build the full term-loan schedule (one row per year until fully repaid).
 * Returns an empty-but-safe zero row when there's nothing to amortise.
 *
 * Mirrors backend/calculations/loan_schedule.py's two CA-reviewer-driven
 * fixes — both of which must live here too, since this file exists
 * specifically to avoid disagreeing with the backend:
 *  1. Moratorium is tracked in HALF-YEAR units (round(months/6)), not
 *     whole years via Math.ceil(months/12) — the old whole-year version
 *     silently DOUBLED a 6-month moratorium to a full 12-month one.
 *  2. The half-yearly instalment is rounded to the nearest WHOLE RUPEE
 *     (not paisa) — the unit everything is actually displayed in — and
 *     the final half-year absorbs whatever residual that rounding
 *     leaves, so principal instalments always sum exactly to the loan
 *     amount instead of landing a few rupees short.
 */
export function buildLoanSchedule(
  loanAmount: number,
  interestRatePct: number,
  tenureMonths: number,
  moratoriumMonths: number = 0,
): LoanScheduleYear[] {
  if (!(loanAmount > 0) || !(tenureMonths > 0)) {
    return [{ year: 1, openingBalance: 0, interestPaid: 0, principalPaid: 0, closingBalance: 0, halfYearlyInstalment: 0 }];
  }

  const halfRate = interestRatePct / 100 / 2;
  const tenureYears = Math.ceil(tenureMonths / 12);
  const totalHalfYears = tenureYears * 2;
  const moratoriumHalfYears = Math.min(Math.round(moratoriumMonths / 6), totalHalfYears);
  const repayHalfYears = Math.max(totalHalfYears - moratoriumHalfYears, 1);
  const halfInst = Math.round(loanAmount / repayHalfYears);

  const rows: LoanScheduleYear[] = [];
  let balance = loanAmount;

  for (let yr = 1; yr <= tenureYears; yr++) {
    const opening = Rs2(balance);
    let bal = opening;

    const hy1 = 2 * yr - 1;
    const ih1 = Rs2(bal * halfRate);
    const repaidH1 = hy1 <= moratoriumHalfYears ? 0 : (hy1 === totalHalfYears ? bal : Math.min(halfInst, bal));
    bal = Rs2(Math.max(bal - repaidH1, 0));

    const hy2 = 2 * yr;
    const ih2 = Rs2(bal * halfRate);
    const repaidH2 = hy2 <= moratoriumHalfYears ? 0 : (hy2 === totalHalfYears ? bal : Math.min(halfInst, bal));
    const closing = Rs2(Math.max(bal - repaidH2, 0));

    const principalPaid = Rs2(repaidH1 + repaidH2);
    const interestPaid = Rs2(ih1 + ih2);

    rows.push({
      year: yr,
      openingBalance: opening,
      interestPaid,
      principalPaid,
      closingBalance: closing,
      halfYearlyInstalment: halfInst,
    });
    balance = closing;
  }

  return rows;
}

/** Year-1 interest + principal — the figure most "quick estimate" UI needs. */
export function getYear1LoanFigures(
  loanAmount: number,
  interestRatePct: number,
  tenureMonths: number,
  moratoriumMonths: number = 0,
): { interestPaid: number; principalPaid: number } {
  const schedule = buildLoanSchedule(loanAmount, interestRatePct, tenureMonths, moratoriumMonths);
  return { interestPaid: schedule[0].interestPaid, principalPaid: schedule[0].principalPaid };
}

/**
 * CA-standard DSCR for a single year:
 *   DSCR = (Cash Accruals + TL Interest) / (TL Principal + TL Interest)
 * Cash Accruals here means PAT + Depreciation (or a reasonable proxy —
 * callers building a quick estimate without a full P&L may pass
 * pre-tax/pre-depreciation operating profit, which is still far closer
 * to correct than omitting the interest add-back entirely).
 */
export function calculateDscr(cashAccruals: number, interestPaid: number, principalPaid: number): number {
  const denominator = principalPaid + interestPaid;
  if (denominator <= 0) return 0;
  return Rs2((cashAccruals + interestPaid) / denominator);
}
